iTunes: Fix playlist folder ordering in macOS importer

fwcd opened the pull request:
A small quality-of-life addition to #11353 that preserves the correct order of playlist folders imported with the macOS iTunes importer.

fwcd committed:
ITunesMacOSImporter: Fix order of playlist tree

Swiftb0y commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -81,10 +81,21 @@ void importMediaItems(NSArray<ITLibMediaItem*>* items) {
-                it != m_playlistChildsByDbId.end() && it.key() == playlistId;
-                ++it) {
-            int childId = it.value();
+        // Find bounds of the child ID range in the multimap
+        auto childIdsBegin = m_playlistChildsByDbId.constFind(playlistId);
+        auto childIdsEnd = childIdsBegin;
+        while (childIdsEnd != m_playlistChildsByDbId.constEnd() &&
+                childIdsEnd.key() == playlistId) {
+            ++childIdsEnd;
+        }
+
+        // QMultiMap stores items with same key in reverse insertion order,
+        // therefore iterate in reverse to get the insertion order again.
+        // See the docs: https://doc.qt.io/qt-5/qmultimap.html#details
+        auto it = childIdsEnd;
+        while (it != childIdsBegin) {
+            --it;
+            int childId = *it;
```
wouldn't this be easier by simply using a `std::reverse_iterator`? Then we don't have to linearly search for the element before the end and then do the looping in reverse. Also, is the `it.key() == playlistId` really necessary? I'd assume that an iterator over a multimap doesn't iterate over unrelated keys. 

```suggestion
        std::for_each(std::make_reverse_iterator(m_playlistChildsByDbId.constEnd()),
            std::make_reverse_iterator(m_playlistChildsByDbId.constFind(playlistId)),
            [](int childId) {....};
```
(this is just a rough outline, I don't know if that would work, I'd highly appreciate if you could experiment. Using `std::ranges::views::reverse` could also be worth looking at.)

fwcd replied:
> Also, is the it.key() == playlistId really necessary?

I'm afraid it is, I've based this on an example [from the docs](https://doc.qt.io/qt-5/qmultimap.html#details) (see the section on STL-style iteration). The reverse iterator sounds interesting, I'll have to look into that (though I'm not sure if it will suffice since the find method only gives us the start of the section, we'd have to seek to the end of the section anyway).

fwcd replied:
> Then we don't have to linearly search for the element before the end and then do the looping in reverse.

Since we have to linearly scan over all the elements anyway, I think traversing them twice shouldn't be an issue.

fwcd replied:
`std::ranges::views::reverse` does look pretty interesting, but was introduced in C++20, can we already use that?

daschuer replied:
Unfortunately this relies on undefined behaviour, because constFind() may find one element in the equal range not necessarily the first.

fwcd replied:
Are you sure? I'm basing this off the documented example

```cpp
QMultiMap<QString, int>::iterator i = map.find("plenty");
while (i != map.end() && i.key() == "plenty") {
    cout << i.value() << Qt::endl;
    ++i;
}
```

If `constFind` would find an arbitrary element instead of the first, it would be nearly useless (or much less intuitive, when would that useful? The user would always have to scan backwards too ib that case)

daschuer replied:
> Are you sure? 

No, I am not sure.. 

It obviously works for Qt5 because QMaps in Qt5 is using a red-black three. And according to the link they seems to guarantee the behaviour for the std::multimap implementations as well. 

The issue is just, that there is no such guarantee. See: https://cplusplus.com/reference/map/multimap/find/

> If constFind would find an arbitrary element instead of the first, it would be nearly useless (or much less intuitive, when would that useful? The user would always have to scan backwards too in that case)

Maybe that is the reason they suggest to use [multimap::equal_range](https://cplusplus.com/multimap::equal_range), where the behaviour is guaranteed. 

But I am not an expert, so I will not stress the issue any further.

daschuer commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -81,10 +81,21 @@ void importMediaItems(NSArray<ITLibMediaItem*>* items) {
+
+        // QMultiMap stores items with same key in reverse insertion order,
+        // therefore iterate in reverse to get the insertion order again.
+        // See the docs: https://doc.qt.io/qt-5/qmultimap.html#details
```
This is unfortunately a Qt5 implementation detail. With Qt6 it depends on the used std:: implementation. 
Does the code rely on it? Than we may consider to force it by `QMultiMap<int, std::vector<int>>`

fwcd replied:
That kind of sidesteps the point of using a multimap, in that case we could just use a regular map or hash (which is fine, my thinking was that the multimap is likely more efficient when working on large playlist trees)

fwcd replied:
The naive implementation would be to just get the equal range as a `QList` or `std::vector` and sort it, which is O(n log n), rather than O(n). But that's maybe still acceptable, probably much clearer and doesn't rely on unnecessarily many details. Wdyt?

Swiftb0y replied:
How many values could there be present for a single key? IIU the code correctly, there are many keys but few values. In that case we should optimize for constant overhead instead. Also, if using a `std::multimap` has advantages, lets just use it directly. It's not like we need to pass it to any other Qt Apis. 
`std::multimap::equal_range` seems perfect for our use case. We also don't need to do the reversing because the standard explicitly dictates that the order of the elements in the range is in insertion order unless specified otherwise explicitly).

fwcd replied:
That very much depends on the user's playlist structure. If a user has a deep tree of folders with few playlists, then yes, but if the user prefers a relatively flat hierarchy there will be lots of values for few keys. `std::multimap` sounds promising, I'll look into that.

Swiftb0y replied:
```cpp
// renamed from m_playlistChildsByDbId
std::pair allChildrenOfParentPlaylist = m_playlistDbIdsByParentDbId.equal_range(playlistId);

std::for_each(allChildrenOfParentPlaylist.first, allChildrenOfParentPlaylist.second, [](std::pair kv) {const int childId = kv.second; ...});

```

daschuer commented on src/library/itunes/itunesmacosimporter.mm:
```diff
@@ -81,10 +81,21 @@ void importMediaItems(NSArray<ITLibMediaItem*>* items) {
-                ++it) {
-            int childId = it.value();
+        // Find bounds of the child ID range in the multimap
+        auto childIdsBegin = m_playlistChildsByDbId.constFind(playlistId);
```
Can we rename m_playlistChildsByDbId to m_playlistDbIdsByParentDbId?

fwcd commented:
I've implemented @Swiftb0y's suggestion to use `std::multimap`, which is much shorter and cleaner thanks! Also I've renamed the multimap to `m_playlistDbIdsByParentDbId` as suggested.

Swiftb0y reviewed (approved):
Thank you. Waiting for @daschuer

daschuer reviewed (approved):
Thank you. Looks good!
